common/defaults.go

```diff
@@ -1397,24 +1447,46 @@ func (u *UpstreamConfig) SetDefaults(defaults *UpstreamConfig) error {
 		if defaults != nil && defaults.Failsafe != nil && len(defaults.Failsafe) > 0 {
 			// Apply defaults to each failsafe config
 			for i, fs := range u.Failsafe {
-				// Find matching default by method/finality
+				// Find matching default - don't inherit if upstream has policies defined
 				var defaultFs *FailsafeConfig
-				for _, dfs := range defaults.Failsafe {
-					// Match method using wildcard (if both are specified)
-					methodMatch := true
-					if dfs.MatchMethod != "" && fs.MatchMethod != "" {
-						methodMatch, _ = WildcardMatch(dfs.MatchMethod, fs.MatchMethod)
-					} else if dfs.MatchMethod != "" || fs.MatchMethod != "" {
-						// If only one has a method specified, they don't match
-						methodMatch = false
-					}
+				hasUpstreamPolicies := fs.Retry != nil || fs.Hedge != nil || fs.CircuitBreaker != nil ||
+					fs.Timeout != nil || fs.Consensus != nil
+
+				if !hasUpstreamPolicies {
+					for _, dfs := range defaults.Failsafe {
+						// Convert legacy fields to matchers for matching
+						fs.convertLegacyMatchers()
+						dfs.convertLegacyMatchers()
 
-					// Match finality (empty array means any finality)
-					finalityMatch := matchFinalities(dfs.MatchFinality, fs.MatchFinality)
+						// Use matcher-based matching for method/finality patterns
+						methodMatch := true
+						finalityMatch := true
 
-					if methodMatch && finalityMatch {
-						defaultFs = dfs
-						break
+						if len(fs.Matchers) > 0 && len(dfs.Matchers) > 0 {
+							methodMatch = false
+							finalityMatch = false
+							for _, fsMatcher := range fs.Matchers {
+								for _, dfsMatcher := range dfs.Matchers {
+									if fsMatcher.Method == "" || dfsMatcher.Method == "" || fsMatcher.Method == "*" || dfsMatcher.Method == "*" {
+										methodMatch = true
+									} else {
+										methodMatch, _ = WildcardMatch(dfsMatcher.Method, fsMatcher.Method)
+									}
+									finalityMatch = matchFinalities(dfsMatcher.Finality, fsMatcher.Finality)
+									if methodMatch && finalityMatch {
+										break
+									}
+								}
+								if methodMatch && finalityMatch {
+									break
+								}
```

Comment:
Should we also apply other matching criteria vs just method/finality? 

I'm not even sure if this while defaulting logic is correct and intuitive, in terms of what the users would expect when defining failsafe policies in the upstreamDefaults + upstreams themselves.

The current logic probably has some unexpected behaviors, as in mixing certain defaults even though user doesn't expect due to criteria difference.

If there's no clean way of doing this "defauling" logic can be that it overrides the whole policies?